hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java

```diff
@@ -720,29 +759,38 @@ private int checkMissingStates() {
     }
 
     /*
-        Invoked when a data is added on a different node and this data should be propagated
+        Data incoming from a different node.  This has already been globally filtered but not locally filtered.
+        It does not need to be re-propagated.
+
+        We allow concurrent threads to make this call in order to process distributed data in parallel. We
+        use synchronized blocks to protect pendingData.
      */
     @Override
+    @Lock(LockType.READ)
     public void onNewData(Data data) {
-        addPendingData(data);
+        sendData(data);
```

Comment:
Arg!  Thanks!  I was tired, I guess.  That is the whole point of having the private "addData" method :)